File: webgasdospela/parser.ts
import * as P from "parsimmon";
import * as lang from "./lang";

var whitespace = P.regexp(/\s*/m);

function token<T>(p: P.Parser<T>): P.Parser<T> {
  return p.skip(whitespace);
}

let ts = (s: string) => token(P.string(s));

var lbrace = ts('{');
var rbrace = ts('}');
var lpar = ts('(');
var rpar = ts(')');
var comma = ts(',');
var colon = ts(':');
var semiColon = ts(';');
var setEq = ts(':=');
var isEq = ts('=');
var notEq = ts('!=');
var then = ts('>>');
var ampersand = ts('&');
var question = ts('?');

function commaSep<T>(parser: P.Parser<T>) {
  return P.sepBy(parser, comma);
}
function semiColonSep<T>(parser: P.Parser<T>) {
  return P.sepBy(parser, semiColon);
}

var stringLiteral =
  token(P.regexp(/"((?:\\.|.)*?)"/, 1))
    .desc('string');

const id = token(P.regexp(/[a-z][a-zA-Z0-9.]*/)).desc("id");
const ctor = token(P.regexp(/[A-Z][a-zA-Z0-9]*/)).desc("ctor").map((s) => new lang.Constant(s));

const exp: P.Parser<lang.Expression> = P.lazy(() =>
  whitespace.then(P.alt(
        id.map((s) => new lang.VariableDeref(s)), ctor
    )));

const comparisonOp =
    P.alt(isEq.result(lang.ComparisonOperator.EQ), notEq.result(lang.ComparisonOperator.NEQ));
const comparison = P.seqMap(exp, comparisonOp, exp, (e1, compOp, e2) => new lang.Condition(e1, compOp, e2))

export const comparisons = commaSep(comparison);

const update = P.seq(id.skip(setEq), ctor).map(([id, c]) => new lang.Update(id, c));

const action = P.seq(id.skip(lpar), commaSep(id).skip(rpar)).map(([id, names]) => new lang.Action(id, names));

const arule = P.seq(action.skip(then).skip(lbrace),
                    semiColonSep(update).skip(rbrace),
                    P.alt(ampersand.skip(lbrace).then(stringLiteral).skip(rbrace).map((s) => new lang.PrintSideEffect(s)), P.succeed(undefined))
                    ).map(([action, updates, sideEffect]) => new lang.ARule(action, updates, sideEffect));

const TabSize = 4;

const prule =
    P.seq(P.index, lbrace.then(comparisons).skip(rbrace).skip(question))
        .chain(([index, conds]) =>
            P.index.chain((i) => i.column === index.column + TabSize ? rule : P.fail("no more")).atLeast(1).map((rules) => new lang.PRule(conds, rules)));

const rule: P.Parser<lang.Rule> = P.lazy(() => P.alt(prule, arule));

const initialState = ts("Initial").skip(lbrace).then(semiColonSep(update)).skip(rbrace).map((updates) => new lang.InitBlock(updates));

const decl = P.alt(initialState, rule)

export const story = decl.many().map(decls => new lang.Story(decls));

File: fsharpAnalysis/GasdospelaToSMV/GasdospelaToSMV/Parser.fs
namespace GasdospelaToSMV

open FParsec
open AST

module GasdospelaParser =

    type Parser<'Result> = Parser<'Result, unit>

    let token (s:string) : Parser<unit> = skipString s  .>> spaces
    let setEq = token ":="
    let isEq = token "="
    let notEq = token "!="
    let comma = token ","
    let semiColon = token ";"
    let lbrace = token "{"
    let rbrace = token "}"
    let lpar = token "("
    let rpar = token ")"
    let question = token "?"
    let amp = token "&"
    let gtgt = token ">>"
    let qt = token "\""

    let initial = token "Initial"

    let isAsciiIdStart c =
        isUpper c || c = '_'

    let isAsciiIdContinue c =
        isAsciiLetter c || isDigit c

    let ctor = identifier (IdentifierOptions(isAsciiIdStart    = isUpper,
                                  isAsciiIdContinue = isAsciiIdContinue)) .>> spaces

    let isAsciiIdContinueVar c =
        isAsciiLetter c || isDigit c || c = '.'

    let var = identifier (IdentifierOptions(isAsciiIdStart    = isLower,
                                  isAsciiIdContinue = isAsciiIdContinueVar)) .>> spaces

    let inPars p = lpar >>. p .>> rpar
    let inBraces p = lbrace >>. p .>> rbrace

    let action = var .>>. (inPars <| sepBy var comma)

    let update: Parser<Update> = var .>> setEq .>>. ctor

    let updates = sepBy update semiColon

    let exp =
        (var |>> VariableDeref)
            <|>
        (ctor |>> Constant)

    let compOp =
        (isEq >>% EQ)
            <|>
        (notEq >>% NEQ)

    let arule: Parser<Rule> =
        tuple3
            (action .>> gtgt)
            (inBraces updates)
            (opt(amp >>. inBraces(qt >>. manySatisfy ((<>)'\"') .>> qt)))
            |>> ARule

    let comparisons = sepBy(tuple3 exp compOp exp) comma

    let rule, ruleRef = createParserForwardedToRef<Rule, unit>()

    let getColumn = getPosition |>> (fun pos -> pos.Column)

    let TabSize = 4L

    let atNextTabStop = getPosition >>= (fun pos -> userStateSatisfies (fun curr -> curr = (pos.Column - TabSize)))

    let prule: Parser<Rule> =
        getColumn .>>. inBraces comparisons .>> question
            >>= (fun(startColumn, comps) ->
                    many1(getColumn >>= (fun col -> if col = startColumn + TabSize then rule else pzero))
                        |>> (fun rules -> PRule(comps, rules)))

    do ruleRef := prule <|> arule

    let initialDecl = initial >>. inBraces updates |>> InitBlock

    let decl = attempt (initialDecl <|> (rule |>> Rule))
    let decls = (spaces >>. many decl .>> eof)

    let parseString str = run decls str
    let parseFile path =
        match runParserOnFile decls () path System.Text.Encoding.UTF8 with
        | Success(s,_,_) -> s
        | Failure(_, f, _) -> failwithf "error: %A" f
